feat: Slinky sidecar integration

Eric-Warehime opened the pull request:
### Changelist
[Describe or list the changes made in this PR]

### Test Plan
[Describe how this PR was tested (if applicable)]

### Author/Reviewer Checklist
- [ ] If this PR has changes that result in a different app state given the same prior state and transaction list, manually add the `state-breaking` label.
- [ ] If the PR has breaking postgres changes to the indexer add the `indexer-postgres-breaking` label.
- [ ] If this PR isn't state-breaking but has changes that modify behavior in `PrepareProposal` or `ProcessProposal`, manually add the label `proposal-breaking`.
- [ ] If this PR is one of many that implement a specific feature, manually label them all `feature:[feature-name]`.
- [ ] If you wish to for mergify-bot to automatically create a PR to backport your change to a release branch, manually add the label `backport/[branch-name]`.
- [ ] Manually add any of the following labels: `refactor`, `chore`, `bug`.

tqin7 commented on protocol/daemons/slinky/client/price_fetcher.go:
```diff
@@ -0,0 +1,144 @@
+		// get the market id for the currency pair
+		id, err := p.marketPairFetcher.GetIDForPair(currencyPair)
+		if err != nil {
+			p.logger.Info("slinky client returned currency pair not found in MarketPairFetcher",
```
for my understanding, this happens if slinky returns a pair that doesn't exist in `x/prices`, right?

Eric-Warehime replied:
Yes, the `Prices` endpoint on slinky returns all pairs that are configured to be fetched. Running with a bunch of extra pairs being fetched on the sidecar returns all of these, but shouldn't be considered a failure.

Potentially making this a debug log is smart since it is quite noisy currently.

tqin7 commented on protocol/daemons/slinky/client/constants.go:
```diff
@@ -0,0 +1,20 @@
+	// SlinkyMarketParamFetchDelay is the frequency at which we query the x/price module to refresh mappings from
+	// currency pair to x/price ID.
+	SlinkyMarketParamFetchDelay = time.Millisecond * 1900
+)
```
curious how we arrived at above values for `SlinkyPriceFetchDelay` and `SlinkyMarketParamFetchDelay`? Have we experimented with how low these values can be?

Eric-Warehime replied:
These are just initial values--I haven't experimented too much with them.
Here's my thoughts on how to think about them:

`SlinkyPriceFetchDelay` can be as low as we want it to be. My initial thought was that it doesn't make sense to set it lower than, however, if we use the `min_exchanges` param as the number of historical slinky prices to keep in the indexPriceCache, maybe it makes sense to hit the sidecar a bunch every block and take the median of those values via x/prices GetValidMarketPriceUpdates method.

For the `MarketParamFetchDelay` I just wanted it to refresh for frequently than we retrieved prices so there would never be a mismatch between the keeper state and the prices returned.

Eric-Warehime commented on protocol/app/app.go:
```diff
@@ -751,25 +763,38 @@ func New(
-			app.RegisterDaemonWithHealthMonitor(app.PriceFeedClient, maxDaemonUnhealthyDuration)
+		// Non-validating full-nodes have no need to run the oracle.
+		if !appFlags.NonValidatingFullNode {
+			if daemonFlags.Price.Enabled {
```
Yep, feature flagging for now so we can switch on/off and remove later.

Eric-Warehime commented on protocol/Dockerfile:
```diff
@@ -41,15 +41,24 @@ RUN --mount=type=cache,target=/root/.cache/go-build \
 
 COPY --from=builder /dydxprotocol/build/dydxprotocold /bin/dydxprotocold
+COPY --from=builder /slinky/build/oracle /bin/slinky
+COPY --from=builder /slinky/config/local/oracle.json /etc/oracle.json
+COPY --from=builder /slinky/config/local/market.json /etc/market.json
```
These are just the default configs in the slinky repo. Probably makes sense to ship them in the v4-chain repo though. I can add them and update.

Eric-Warehime commented on protocol/daemons/slinky/client/price_fetcher.go:
```diff
@@ -0,0 +1,144 @@
+// This is because the sidecar pre-aggregates market data.
+func (p *PriceFetcherImpl) FetchPrices(ctx context.Context) error {
+	// get prices from slinky sidecar via GRPC
+	slinkyResponse, err := p.slinky.Prices(ctx, &types.QueryPricesRequest{})
```
https://github.com/dydxprotocol/v4-chain/pull/1109/files/88675f5198c9cea74428260f03142cc32568c462#diff-a9a1c1853375ca2e04b66ac0adca22f0b701a3c8f1d6325a7f4a277e3453c184R92-R93

The slinky Client uses `ReportFailure` on the HealthCheckable interface and logs an error. During sidecar downtime this daemon would report as unhealthy and no prices would be added to VEs.

Eric-Warehime commented on protocol/daemons/slinky/client/price_fetcher.go:
```diff
@@ -0,0 +1,144 @@
+	var updates []*api.MarketPriceUpdate
+	for currencyPairString, priceString := range slinkyResponse.Prices {
+		// convert currency-pair string (index) into currency-pair object
+		currencyPair, err := oracletypes.CurrencyPairFromString(currencyPairString)
```
Yes. It probably makes sense for the delimiter to be configurable. I can make that change on the slinky side and input `-` for parity w/ existing IDs.

Eric-Warehime commented:
Added some changes:
1. The sidecar daemon now writes directly to the price keeper's index price cache instead of using the pricefeed server.
2. I've introduced a new constant which represents the number of sidecar prices which will be kept in cache. The median of these will be what is proposed for inclusion in blocks, and we can tune the freshness of that data by increasing the frequency at which we poll the sidecar for prices.

The idea behind 2. is to make this backwards compatible with existing min_exchange params already in state. It also has nice properties around further mitigating outliers that might arise.
